Temporal: Make ExactTime::difference() return an InternalDuration

catamorphism opened the pull request:
#### 4927ce289367605b71e2599379961f52ccb31b4c
<pre>
Temporal: Make ExactTime::difference() return an InternalDuration

<a href="https://bugs.webkit.org/show_bug.cgi?id=223166">https://bugs.webkit.org/show_bug.cgi?id=223166</a>

Reviewed by Keith Miller.

TemporalInstant::difference() now calls ExactTime::difference(),
then converts the result back to a Duration.

This enables some of the tests for TemporalInstant/p/since and
TemporalInstant/p/until that previously failed.

As part of this change, implement ExactTime::round() according to the
spec.

* JSTests/stress/temporal-instant.js:
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/ISO8601.cpp:
(JSC::ISO8601::roundTemporalInstant):
(JSC::ISO8601::validateTemporalRoundingIncrement):
(JSC::ISO8601::ExactTime::round const):
(JSC::ISO8601::roundTimeDurationToIncrement):
(JSC::ISO8601::roundTimeDuration):
(JSC::ISO8601::ExactTime::difference const):
(JSC::ISO8601::absInt128): Deleted.
(JSC::ISO8601::ExactTime::round): Deleted.
* Source/JavaScriptCore/runtime/ISO8601.h:
(JSC::ISO8601::ExactTime::absInt128):
* Source/JavaScriptCore/runtime/TemporalDuration.cpp:
(JSC::TemporalDuration::temporalDurationFromInternal):
* Source/JavaScriptCore/runtime/TemporalDuration.h:
* Source/JavaScriptCore/runtime/TemporalInstant.cpp:
* Source/JavaScriptCore/runtime/TemporalObject.cpp:
(JSC::applyUnsignedRoundingMode):
(JSC::roundNumberToIncrementAsIfPositive):
(JSC::roundNumberToIncrementInt128):
* Source/JavaScriptCore/runtime/TemporalObject.h:
(JSC::lengthInNanoseconds):
(JSC::getUnsignedRoundingMode):

Canonical link: <a href="https://commits.webkit.org/299050@main">https://commits.webkit.org/299050@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0622f719a5b32108ca1553ad3b5db6986fad9b70

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/117587 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/37264 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/27889 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/123699 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/69594 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/37957 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/45847 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89235 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/45041 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120539 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30211 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105428 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/69737 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29273 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/23544 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67368 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/109697 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99627 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/23726 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/126812 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/116096 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44490 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33469 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97900 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/44848 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101657 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97688 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24870 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43059 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21013 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/40848 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/44361 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/50036 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/144794 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/43819 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/37262 "Found 1 new JSC binary failure: testapi, Found 20047 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Array/array_slice.js.default, ChakraCore.yaml/ChakraCore/test/Array/array_sort2.js.default, ChakraCore.yaml/ChakraCore/test/Array/push2.js.default, ChakraCore.yaml/ChakraCore/test/Array/reverse1.js.default, ChakraCore.yaml/ChakraCore/test/Array/toLocaleString.js.default, ChakraCore.yaml/ChakraCore/test/Bugs/blue_1086262.js.default, ChakraCore.yaml/ChakraCore/test/Bugs/bug215238.shrua-2.js.default, ChakraCore.yaml/ChakraCore/test/ControlFlow/forInPrimitive.js.default, ChakraCore.yaml/ChakraCore/test/Date/date_cache_bug.js.default, ChakraCore.yaml/ChakraCore/test/Error/CallNonFunction.js.default ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/47167 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45511 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/ISO8601.cpp:
```diff
@@ -1739,15 +1797,15 @@ InternalDuration InternalDuration::combineDateAndTimeDuration(Duration dateDurat
-    return round(diff, increment, unit, roundingMode);
-}
+    VM& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
```
This is not necessary.

catamorphism replied:
I think it's necessary, because `ExactTime::difference()` calls `roundTimeDuration()`, which can throw.

catamorphism replied:
Never mind, I misunderstood. Fixed in [b6b5887](https://github.com/WebKit/WebKit/pull/45603/commits/b6b5887b72e06d0f96be84f28e79e0377f1e4a1e)

Constellation commented on Source/JavaScriptCore/runtime/ISO8601.cpp:
```diff
@@ -1739,15 +1797,15 @@ InternalDuration InternalDuration::combineDateAndTimeDuration(Duration dateDurat
-    return ExactTime { round(m_epochNanoseconds, increment, unit, roundingMode) };
+    Int128 timeDuration = other.m_epochNanoseconds - m_epochNanoseconds;
+    timeDuration = roundTimeDuration(globalObject, timeDuration, roundingIncrement, smallestUnit, roundingMode);
+    RETURN_IF_EXCEPTION(scope, { });
```
Not necessary.

catamorphism replied:
I think it is? `roundTimeDuration()` can throw.

catamorphism replied:
Sorry, fixed in [b6b5887](https://github.com/WebKit/WebKit/pull/45603/commits/b6b5887b72e06d0f96be84f28e79e0377f1e4a1e)

Constellation commented on Source/JavaScriptCore/runtime/ISO8601.cpp:
```diff
@@ -1680,20 +1674,84 @@ std::optional<ExactTime> ExactTime::add(Duration duration) const
+    TemporalUnit unit, RoundingMode roundingMode)
+{
+    VM& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
```
Not necessary.

catamorphism replied:
Fixed in [b6b5887](https://github.com/WebKit/WebKit/pull/45603/commits/b6b5887b72e06d0f96be84f28e79e0377f1e4a1e)

Constellation commented on Source/JavaScriptCore/runtime/ISO8601.cpp:
```diff
@@ -1680,20 +1674,84 @@ std::optional<ExactTime> ExactTime::add(Duration duration) const
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    auto divisor = lengthInNanoseconds(unit);
+    RELEASE_AND_RETURN(scope, roundTimeDurationToIncrement(globalObject, timeDuration,
```
Just use `return roundTimeDurationToIncrement(...)`.

catamorphism replied:
I thought it had to be `RELEASE_AND_RETURN(...)` because `roundTimeDurationToIncrement()` can throw?

ptomato replied:
I'm not sure if this is what @Constellation is referring to, but maybe the tail-call style would just use `roundTimeDurationToIncrement`'s throw scope and creating an outer one wouldn't be necessary, since this is the only thing in the scope that can throw? Just a guess, I don't know off the top of my head if that's accurate.

catamorphism replied:
@ptomato Ah, that's exactly it. Fixed in [b6b5887](https://github.com/WebKit/WebKit/pull/45603/commits/b6b5887b72e06d0f96be84f28e79e0377f1e4a1e)

ptomato commented on JSTests/stress/temporal-instant.js:
```diff
@@ -427,3 +427,19 @@ const maxValue = new Temporal.Instant(86400_0000_0000_000_000_000n);
         shouldThrow(() => epoch.subtract({ [unit]: -Number.MAX_VALUE }), RangeError);
     });
 }
+
+// since() and until()
+{
+    const earlier = Temporal.Instant.from("1976-11-18T15:23:30.123456789Z");
+    const later = Temporal.Instant.from("2019-10-29T10:46:38.271986102Z");
+    const diff = later.since(earlier);
+    shouldBe(diff.nanoseconds, 313);
+    shouldBe(diff.microseconds, 529);
+    shouldBe(diff.milliseconds, 148);
+    shouldBe(diff.seconds, 1355167388);
+    const diff1 = earlier.until(later);
+    shouldBe(diff1.nanoseconds, diff.nanoseconds);
+    shouldBe(diff1.microseconds, diff.microseconds);
+    shouldBe(diff1.milliseconds, diff.milliseconds);
+    shouldBe(diff1.seconds, diff.seconds);
+}
```
Optional: I think this is probably sufficiently covered by test262 and doesn't need an additional stress test; can't hurt though

catamorphism replied:
I thought that all changes were supposed to have stress tests, but I agree it does overlap with test262.

ptomato commented on Source/JavaScriptCore/runtime/ISO8601.cpp:
```diff
@@ -1680,20 +1674,84 @@ std::optional<ExactTime> ExactTime::add(Duration duration) const
+
+// https://tc39.es/proposal-temporal/#sec-validatetemporalroundingincrement
+static void validateTemporalRoundingIncrement(JSGlobalObject* globalObject, unsigned increment,
+    Int128 dividend, bool inclusive)
```
On other PRs I've gotten the feedback that something like `enum { Inclusive, NotInclusive }` is preferred instead of a boolean. Not sure if that's the case here.

catamorphism replied:
Changed in [1a7478f](https://github.com/WebKit/WebKit/pull/45603/commits/1a7478ffedbc22e97ccd46a002d9b3f7cc8072a9)

ptomato commented on Source/JavaScriptCore/runtime/ISO8601.cpp:
```diff
@@ -1680,20 +1674,84 @@ std::optional<ExactTime> ExactTime::add(Duration duration) const
+
+    Int128 rounded = roundNumberToIncrementInt128(d, increment, roundingMode);
+    if (absInt128(rounded) > InternalDuration::maxTimeDuration) {
+        throwRangeError(globalObject, scope, "Rounded instant exceeds maximum time duration"_s);
```
```suggestion
        throwRangeError(globalObject, scope, "Rounded time duration exceeds maximum"_s);
```
(Or a more descriptive error message based on where this is called from, but that could be difficult as it's called from several places, not all of which correspond to a time duration that the user supplied)

catamorphism replied:
Fixed the error message in [c719f90](https://github.com/WebKit/WebKit/pull/45603/commits/c719f90b74999ab3f185aa9f6f54ad72d3749c37) . I agree that a more descriptive error message could be difficult.

ptomato commented on Source/JavaScriptCore/runtime/TemporalDuration.cpp:
```diff
@@ -378,6 +378,92 @@ ISO8601::Duration TemporalDuration::add(JSGlobalObject* globalObject, JSValue ot
+        microseconds = microseconds % 1000;
+    } else if (largestUnit == TemporalUnit::Microsecond) {
+        microseconds = (Int128) std::trunc(std::fma<double, unsigned, double>(
+            seconds, 6, std::trunc(((double) nanoseconds) / 1000)));
```
I think 6 is probably meant to be 10⁶? If so, this should make more test262 tests pass (and if it doesn't, we should add coverage for this)

catamorphism replied:
Good catch! It turns out the code was correct because `seconds` is always 0. (I got there from following the code in the polyfill, but in the polyfill time durations are represented with seconds and subseconds separately, and in JSC it's a single nanoseconds value.) But the call to `std::fma` is unnecessary so I changed it in [c719f90](https://github.com/WebKit/WebKit/pull/45603/commits/c719f90b74999ab3f185aa9f6f54ad72d3749c37) to be similar to the other cases.

ptomato commented on Source/JavaScriptCore/runtime/TemporalObject.cpp:
```diff
@@ -600,55 +627,78 @@ double roundNumberToIncrement(double x, double increment, RoundingMode mode)
-    if (increment == 1)
-        return x;
-
+// The following code follows the polyfill rather than the spec, because we don't have float128.
```
I would suggest rephrasing this comment and the one below not to refer to "the polyfill", it's not clear from context here what is meant.

catamorphism replied:
I added links to the relevant code in the polyfill in [86a28f8](https://github.com/WebKit/WebKit/pull/45603/commits/86a28f87d22a6bd6f24f140318c2fd07be727fdd) -- hopefully that clarifies?

ptomato reviewed (commented):
I'm not a reviewer, but I had some suggestions that will hopefully help the actual review.

kmiller68 commented on Source/JavaScriptCore/runtime/ISO8601.cpp:
```diff
@@ -1680,20 +1674,85 @@ std::optional<ExactTime> ExactTime::add(Duration duration) const
+    case TemporalUnit::Minute: maximum = (Int128) (minutesPerHour * WTF::hoursPerDay); break;
+    case TemporalUnit::Second: maximum = (Int128) (secondsPerMinute * minutesPerHour * WTF::hoursPerDay); break;
+    case TemporalUnit::Millisecond: maximum = (Int128) msPerDay; break;
+    case TemporalUnit::Microsecond: maximum = (Int128) msPerDay * 1000; break;
```
You shouldn't need the `(Int128)` and if for some reason it is needed WebKit style would be to `static_cast<Int128>` rather than C-style cast.

catamorphism replied:
I think the cast was necessary to get it to compile on armv7, but I'll try it without.

catamorphism replied:
Yup, armv7 fails if I remove the casts; I re-added them in the `static_cast` form.

kmiller68 commented on Source/JavaScriptCore/runtime/TemporalObject.h:
```diff
@@ -114,6 +122,11 @@ enum class UnitGroup : uint8_t {
     Time,
 };
 
+enum IsInclusive : bool {
```
I would make this: `enum class Inclusitivity : bool {`. `Inclusive` and `Exclusive` seem like common enough words I wouldn't want to clobber the global namespace. Also, the `class` gives some type safety.

catamorphism replied:
Done in [9ba1c59](https://github.com/WebKit/WebKit/pull/45603/commits/9ba1c59d52b0b620796394042548a8b89c644855)

kmiller68 commented on Source/JavaScriptCore/runtime/TemporalObject.h:
```diff
@@ -132,8 +145,60 @@ std::optional<double> maximumRoundingIncrement(TemporalUnit);
+    case TemporalUnit::Microsecond:
+        return 1000;
+    case TemporalUnit::Millisecond:
+        return 1000000;
```
I would do these as `1 << X`, `1000 * 1000`, or `1000 * lengthInNanoseconds(TemporalUnit::Microsecond);` Those are a bit easier to read and the compiler will fold it all away.

catamorphism replied:
Done in [9ba1c59](https://github.com/WebKit/WebKit/pull/45603/commits/9ba1c59d52b0b620796394042548a8b89c644855)

kmiller68 commented on Source/JavaScriptCore/runtime/TemporalObject.cpp:
```diff
@@ -600,55 +627,80 @@ double roundNumberToIncrement(double x, double increment, RoundingMode mode)
+{
+// This follows the polyfill code rather than the spec, in order to work around
+// being unable to apply floating-point division in x / increment.
+// See https://github.com/tc39/proposal-temporal/blob/main/polyfill/lib/ecmascript.mjs#L4043
```
This should be aligned with the code.

catamorphism replied:
Done in [9ba1c59](https://github.com/WebKit/WebKit/pull/45603/commits/9ba1c59d52b0b620796394042548a8b89c644855)

kmiller68 commented on Source/JavaScriptCore/runtime/TemporalObject.cpp:
```diff
@@ -600,55 +627,80 @@ double roundNumberToIncrement(double x, double increment, RoundingMode mode)
+// The following code follows the polyfill rather than the spec, because we don't have float128.
+// ApplyUnsignedRoundingMode is inlined here to mirror the polyfill's implementation of it,
+// which has a different type than in the spec.
+// See https://github.com/tc39/proposal-temporal/blob/main/polyfill/lib/ecmascript.mjs#L4056
```
Ditto on spacing.

catamorphism replied:
Done in [9ba1c59](https://github.com/WebKit/WebKit/pull/45603/commits/9ba1c59d52b0b620796394042548a8b89c644855)

kmiller68 commented on Source/JavaScriptCore/runtime/TemporalObject.cpp:
```diff
@@ -600,55 +627,80 @@ double roundNumberToIncrement(double x, double increment, RoundingMode mode)
+        return r1 * increment;
+    if (unsignedRoundingMode == UnsignedRoundingMode::Infinity)
+        return r2 * increment;
+    if (cmp < 0)
```
Let's make this `if (doubleRemainder < increment)`

catamorphism replied:
Done in [9ba1c59](https://github.com/WebKit/WebKit/pull/45603/commits/9ba1c59d52b0b620796394042548a8b89c644855)

kmiller68 commented on Source/JavaScriptCore/runtime/TemporalObject.cpp:
```diff
@@ -600,55 +627,80 @@ double roundNumberToIncrement(double x, double increment, RoundingMode mode)
+        return r2 * increment;
+    if (cmp < 0)
+        return r1 * increment;
+    if (cmp > 0)
```
Let's make this `if (doubleRemainder > increment)`

catamorphism replied:
Done in [9ba1c59](https://github.com/WebKit/WebKit/pull/45603/commits/9ba1c59d52b0b620796394042548a8b89c644855)

kmiller68 commented on Source/JavaScriptCore/runtime/TemporalObject.cpp:
```diff
@@ -600,55 +627,80 @@ double roundNumberToIncrement(double x, double increment, RoundingMode mode)
+        rounded = r1;
+    else if (unsignedRoundingMode == UnsignedRoundingMode::Infinity)
+        rounded = r2;
+    else if (cmp < 0)
```
Let's make this `if (absInt128(remainder * 2) < increment)`

catamorphism replied:
Done in [9ba1c59](https://github.com/WebKit/WebKit/pull/45603/commits/9ba1c59d52b0b620796394042548a8b89c644855)

kmiller68 commented on Source/JavaScriptCore/runtime/TemporalObject.cpp:
```diff
@@ -600,55 +627,80 @@ double roundNumberToIncrement(double x, double increment, RoundingMode mode)
+        rounded = r2;
+    else if (cmp < 0)
+        rounded = r1;
+    else if (cmp > 0)
```
Let's make this `if (absInt128(remainder * 2) > increment)`

catamorphism replied:
Done in [9ba1c59](https://github.com/WebKit/WebKit/pull/45603/commits/9ba1c59d52b0b620796394042548a8b89c644855)

kmiller68 reviewed (approved):
r=me

catamorphism committed:
Temporal: Make ExactTime::difference() return an InternalDuration

https://bugs.webkit.org/show_bug.cgi?id=223166

Reviewed by Keith Miller.

TemporalInstant::difference() now calls ExactTime::difference(),
then converts the result back to a Duration.

This enables some of the tests for TemporalInstant/p/since and
TemporalInstant/p/until that previously failed.

As part of this change, implement ExactTime::round() according to the
spec.

* JSTests/stress/temporal-instant.js:
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/ISO8601.cpp:
(JSC::ISO8601::roundTemporalInstant):
(JSC::ISO8601::validateTemporalRoundingIncrement):
(JSC::ISO8601::ExactTime::round const):
(JSC::ISO8601::roundTimeDurationToIncrement):
(JSC::ISO8601::roundTimeDuration):
(JSC::ISO8601::ExactTime::difference const):
(JSC::ISO8601::absInt128): Deleted.
(JSC::ISO8601::ExactTime::round): Deleted.
* Source/JavaScriptCore/runtime/ISO8601.h:
(JSC::ISO8601::ExactTime::absInt128):
* Source/JavaScriptCore/runtime/TemporalDuration.cpp:
(JSC::TemporalDuration::temporalDurationFromInternal):
* Source/JavaScriptCore/runtime/TemporalDuration.h:
* Source/JavaScriptCore/runtime/TemporalInstant.cpp:
* Source/JavaScriptCore/runtime/TemporalObject.cpp:
(JSC::applyUnsignedRoundingMode):
(JSC::roundNumberToIncrementAsIfPositive):
(JSC::roundNumberToIncrementInt128):
* Source/JavaScriptCore/runtime/TemporalObject.h:
(JSC::lengthInNanoseconds):
(JSC::getUnsignedRoundingMode):

Canonical link: https://commits.webkit.org/299050@main
